### PR TITLE
deploy: configure `sd-whonix` by package

### LINKED ISSUE #936: [4.2] Write vm-specific config values to qubesdb
Tracking/discussion. Let's make use of `.vm-config` part of qubesdb (internal to a VM) to store vm-specific configuration. Eventually this will allow us to get rid of config.json.

Upstream documentation:
https://dev.qubes-os.org/projects/core-admin-client/en/latest/manpages/qvm-features.html#vm-config

## Configurable components

| Status | Value | `dom0`[1] | `sd-app` | `sd-gpg` | `sd-log` | `sd-proxy` | `sd-whonix` |
| --- | --- | --- | --- | --- | --- | --- | --- |
| Drafted | `$QUBES_GPG_DOMAIN` | | [`$QUBES_GPG_DOMAIN`](https://github.com/freedomofpress/securedrop-client/blob/520e6a7490ebd0f1580b52e40127cc913c55e646/client/files/sd-app-qubes-gpg-domain.sh)[2] |
| n/a | `config.json:.environment` | →`/etc/apt/sources.list.d/securedrop_workstation.list` |
| Drafted | `config.json:.hidserv.hostname` | | | | | `/home/user/.securedrop_proxy/sd-proxy.yaml`[2] | `/var/lib/tor/keys/app_journalist.auth_private`[3] |
| Drafted | `config.json:.hidserv.key` | | | | | | `/var/lib/tor/keys/app_journalist.auth_private`[3] |
| Drafted | `config.json:.submission_key_fpr` | | `/home/user/.securedrop_client/config.json`[2] |
| n/a | `config.json:.vmsizes.sd_app` | `qvm-volume resize sd-app:private ...` |
| n/a | `config.json:.vmsizes.sd_log` | `qvm-volume resize sd-log:private ...` |
| | `salt://sd/sd-journalist.sec` | | | `/home/user/.gnupg/sd-journalist.sec`[4] |

1. Out of scope of this ticket; just documenting here for completeness.
2. [x] Program reads directly
3. [x] Helper program templates into file contents
4. [x] **Secret:** install directly into domain and remove from `dom0`

**NB.**  Only testable on Qubes 4.2, where QubesDB starts to return `vm-config.*` as `/vm-config/*`.

## Related
- https://github.com/freedomofpress/securedrop-workstation/issues/853

### LINKED ISSUE #1039: provide reusable script for rendering templates from QubesDB variables
## Description

`securedrop-workstation` should provide a reusable script that VMs can use to template configuration files from QubesDB for use by third-party programs.
* Prototype: https://github.com/freedomofpress/securedrop-client/blob/27cc69d7a87c937ff31bae9190b68ee2138525b8/with-qubesdb

## How will this impact [SecureDrop/SecureDrop Workstation users](https://github.com/freedomofpress/securedrop-ux/wiki/Users)?
No user-facing changes.

## How would this affect the SecureDrop Workstation [threat model](https://github.com/freedomofpress/securedrop-workstation/#threat-model)?
Discussed in #936 and #1013.

## User Stories
- #1004
- #1029

### PR DESCRIPTION
**Pull request by cfm**
## Status

Ready for review
- ~~Blocked on freedomofpress/securedrop-client#2032~~

## Description of Changes

Closes #1039 by configuring `sd-whonix` using freedomofpress/securedrop-client#2032's `securedrop-whonix-config` rather than Salt.

Closes #936.

## Testing

If testing after freedomofpress/securedrop-client#2032:

1. `make clone && make dev`

...and if testing in conjunction with freedomofpress/securedrop-client#2032:

2. On failure in `sd-base-bookworm-template`: manually install `securedrop-qubesdb`
3. `make dev` again
4. On failure in `whonix-gateway-17`: manually install `securedrop-qubesdb` and `securedrop-whonix-config`
5. `make dev again`

Either way, finally:

6. [ ] You're able to log into the Client.

### If you have made changes to the provisioning logic

- [x] All tests (`make test`) pass in `dom0`

### If you have added or removed files

- [x] I have updated `MANIFEST.in` and `rpm-build/SPECS/securedrop-workstation-dom0-config.spec`

### CODE REVIEW

File: dom0/sd-workstation.top
Comment (cfm) [inline dom0/sd-workstation.top:49]: By strict analogy with the Debian templates, this is our cue to clone `whonix-gateway-17` to something like `sd-whonix-gateway-17` and use that as `sd-whonix`'s template, rather than installing our packages directly into `whonix-gateway-17`.  @legoktm, I know you anticipated this in <https://github.com/freedomofpress/securedrop-workstation/pull/1040#issuecomment-2127551263>, but I'm highlighting it explicitly as a change that should be pretty quick to tack on if we decide it's worth being strict about.

File: dom0/sd-base-template-files.sls
Comment (cfm) [inline dom0/sd-base-template-files.sls:24]: This and the `securedrop-whonix-config` package will fail in CI until freedomofpress/securedrop-client#2032 is merged.

**Comment by legoktm** [comment]
The debs are now on apt-test so I've rebased to kick off CI.

File: dom0/sd-workstation.top
Comment (legoktm) [reply dom0/sd-workstation.top:49]: I'm still trying to reason my way through the implications of cloning the template versus not. My current perspective is that we've already been not cloning it, so we're OK to stick with it. It should be even easier to clean up after ourselves now that everything is packaged.

**Comment by legoktm** [comment]
Do we need a cleanup stage to uninstall the packages? Admittedly I can't really see anything that was cleaning it up before...

**Review by legoktm — CHANGES_REQUESTED** [review]
Running `make dev` doesn't invoke the newly added salt thing, I had to manually run `sudo qubesctl --show-output --skip-dom0 --targets whonix-gateway-17 state.highstate` and then it all worked (and I could log in and auth_private was provisioned correctly).

I think this is because the template is not tagged as `sd-workstation` so it needs an explicit salt run thing.

We also need a salt state to remove the packages (purge ideally) to restore it to original state on uninstall.

File: dom0/sd-clean-all.sls
Comment (legoktm) [inline dom0/sd-clean-all.sls:80]: This doesn't purge the apt-test file if it was created.

File: files/provision-all
Comment (legoktm) [inline files/provision-all:36]: I would appreciate input from a salt guru like @rocodes or @deeplow that this is safe to run (and also the correct invocation).

Comment (rocodes) [reply files/provision-all:36]: This is the correct target, as long as there's an already-created vm present by that name. It does make me want us to clone the whonix template and do all our own operations on the cloned template though, which I think you mentioned somewhere else Kunal, since `provision-all` isn't just run once but any time a migration happens.

As deeplow mentioned this will run whatever states are present/enabled on the machine for the vm by this name.  If a user created a custom .top and .sls file that also targets whonix-gateway-17, that would also be run.  (That's kind of just the symptom of the underlying "should we rely on system templates in an ongoing way or not?" question).

Comment (rocodes) [reply files/provision-all:36]: We've gone back and forth on our whonix provisioning approach in sdw: we previously used `qvm.anon-whonix`, which are upstream tested Salt states that configure the latest whonix vms and set up an up to date disposable anon-whonix vm. Now we invoke the template name directly, needing a little more attention when upgrading, to make sure we upgrade everywhere consistently. (It also doesn't change the user's Tor VMs, which is good or bad depending on how heavy-handed we think we should be).

I kind of think we should:
* run the upstream `qvm.anon-whonix` state again as part of our use of whonix (they describe it as a mandatory step - it configures [update policy information](https://www.whonix.org/wiki/Dev/Qubes#salt), in addition to setting up vms).
* Then clone the whonix gateway and name it whatever we want, and apply the custom configuration to that.

(I think what I'm suggesting can/should be a separate PR)

File: dom0/sd-clean-all.sls
Comment (cfm) [reply dom0/sd-clean-all.sls:80]: Thanks for catching this.  I'm fine with the fix you've proposed in 3114359706310b306a2674fd4cb1a4456bd0f377 until we do what @rocodes proposes in <https://github.com/freedomofpress/securedrop-workstation/pull/1051#discussion_r1624796502>, about which I'll follow up there.

File: files/provision-all
Comment (cfm) [reply files/provision-all:36]: Thanks, all.  @rocodes, if you're happy with your proposal coming as a follow-up change, I'll tackle that next, rather than blocking this.

**Comment by cfm** [comment]
Thanks, all.  Let me tackle the rebase after #1048 and then I'll turn this back over to you, unless you'd rather I address <https://github.com/freedomofpress/securedrop-workstation/pull/1051#discussion_r1624796502> at the same time.

File: dom0/sd-base-template-files.sls
Comment (cfm) [inline dom0/sd-base-template-files.sls:24]: Pending freedomofpress/securedrop-client#2051:

```suggestion
      - securedrop-qubesdb-tools
```

File: dom0/sd-clean-all.sls
Comment (cfm) [inline dom0/sd-clean-all.sls:80]: Pending freedomofpress/securedrop-client#2051:

```suggestion
      - qvm-run whonix-gateway-17 'sudo apt purge --yes securedrop-keyring securedrop-qubesdb-tools securedrop-whonix-config'
```

**Comment by legoktm** [comment]
> and then I'll turn this back over to you, unless you'd rather I address https://github.com/freedomofpress/securedrop-workstation/pull/1051#discussion_r1624796502 at the same time.

No, I double checked with Ro as well and we can defer it to after 4.2 (but before GA).

**Comment by cfm** [comment]
Rebased from `main` after #1048.

**Review by legoktm — APPROVED** [review]
LGTM, just waiting on a green check from SDW CI.

**Comment by legoktm** [comment]
```
INFO:[2024-06-03-23:00:03:129788]                     E: Release file for tor+https://fasttrack.debian.net/debian/dists/bookworm-fasttrack/InRelease is not valid yet (invalid for another 10s). Updates for this repository will not be applied.
```

let's kick off CI again...

**Comment by cfm** [comment]
<https://ws-ci-runner.securedrop.org/2024-06-03-233933015542-5628be940faa556ca8480ec1c7d8715289c0ba61-Qubes_4.2_B-update_20240603044409.log.txt>:

```
INFO:[2024-06-04-00:08:49:204996]             ID: update-apt-cache-with-stable-change
INFO:[2024-06-04-00:08:49:205027]       Function: cmd.run
INFO:[2024-06-04-00:08:49:205055]           Name: apt-get update --allow-releaseinfo-change
INFO:[2024-06-04-00:08:49:205092]         Result: False
INFO:[2024-06-04-00:08:49:205123]        Comment: Command "apt-get update --allow-releaseinfo-change" run
INFO:[2024-06-04-00:08:49:205152]        Started: 00:06:47.633926
INFO:[2024-06-04-00:08:49:205185]       Duration: 1199.845 ms
INFO:[2024-06-04-00:08:49:205213]        Changes:   
INFO:[2024-06-04-00:08:49:205255]                 ----------
INFO:[2024-06-04-00:08:49:205289]                 pid:
INFO:[2024-06-04-00:08:49:205317]                     1823
INFO:[2024-06-04-00:08:49:205349]                 retcode:
INFO:[2024-06-04-00:08:49:205382]                     100
INFO:[2024-06-04-00:08:49:205410]                 stderr:
INFO:[2024-06-04-00:08:49:205442]                     E: Could not get lock /var/lib/apt/lists/lock. It is held by process 1578 (apt-get)
INFO:[2024-06-04-00:08:49:205475]                     E: Unable to lock directory /var/lib/apt/lists/
INFO:[2024-06-04-00:08:49:205507]                 stdout:
INFO:[2024-06-04-00:08:49:205539]                     Reading package lists...
```

File: files/provision-all
Comment (cfm) [reply files/provision-all:36]: Filed as #1060.

**Comment by cfm** [comment]
Hmm.  @legoktm, in the [latest](https://ws-ci-runner.securedrop.org/2024-06-04-001402899146-de4ac9a2a0b25a0e8faeadd71e477c9094667d25-Qubes_4.2_A-update_20240603043744.log.txt) CI failure you flagged:

```
INFO:[2024-06-04-00:44:18:944909] ======================================================================
INFO:[2024-06-04-00:44:18:944941] ERROR: test_v3_auth_private_file (test_sd_whonix.SD_Whonix_Tests.test_v3_auth_private_file)
INFO:[2024-06-04-00:44:18:944972] ----------------------------------------------------------------------
INFO:[2024-06-04-00:44:18:945006] Traceback (most recent call last):
INFO:[2024-06-04-00:44:18:945038]   File "/home/user/securedrop-workstation/tests/test_sd_whonix.py", line 24, in test_v3_auth_private_file
INFO:[2024-06-04-00:44:18:945071]     self.assertFileHasLine("/var/lib/tor/authdir/app-journalist.auth_private", line)
INFO:[2024-06-04-00:44:18:945099]   File "/home/user/securedrop-workstation/tests/base.py", line 139, in assertFileHasLine
INFO:[2024-06-04-00:44:18:945133]     remote_content = self._get_file_contents(remote_path)
INFO:[2024-06-04-00:44:18:945165]                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-06-04-00:44:18:945202]   File "/home/user/securedrop-workstation/tests/base.py", line 98, in _get_file_contents
INFO:[2024-06-04-00:44:18:945251]     return subprocess.check_output(cmd).decode("utf-8")
INFO:[2024-06-04-00:44:18:945287]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-06-04-00:44:18:945321]   File "/usr/lib64/python3.11/subprocess.py", line 466, in check_output
INFO:[2024-06-04-00:44:18:945354]     return run(*popenargs, stdout=PIPE, timeout=timeout, check=True,
INFO:[2024-06-04-00:44:18:945391]            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
INFO:[2024-06-04-00:44:18:945424]   File "/usr/lib64/python3.11/subprocess.py", line 571, in run
INFO:[2024-06-04-00:44:18:945452]     raise CalledProcessError(retcode, process.args,
INFO:[2024-06-04-00:44:18:945487] subprocess.CalledProcessError: Command '['qvm-run', '-p', 'sd-whonix', 'sudo /bin/cat /var/lib/tor/authdir/app-journalist.auth_private']' returned non-zero exit status 1.
INFO:[2024-06-04-00:44:18:945517]
```

I've tacked on 9835669000b3d372c966992b1ed29479e56c0127 to test the "Qubes feature to QubesDB variable" path more rigorously.  But I also see one of our rebases lost the `bash -c` from the [`securedro-whonix-config` service unit's `ExecStartPost`](https://github.com/freedomofpress/securedrop-client/blob/51992805b6d109371616dffb09b3ac0c42706500/debian/securedrop-whonix-config.securedrop-whonix-config.service#L11), which (as I learned and thought I had fixed last week!) doesn't run in a shell.  I'll push that fix up now, but I won't be able to test it locally until tomorrow.

File: files/provision-all
Comment (rocodes) [reply files/provision-all:36]: Thanks, sorry I missed your comment yesterday, but yes, not intended to block this PR, all followup work.

**Comment by cfm** [comment]
freedomofpress/securedrop-client#2056 may help here; with it I'm not able to reproduce this failure locally.  However, Workstation CI does not appear to be running on new commits....

**Comment by legoktm** [comment]
Since CI isn't cooperating I'm just going to verify it manually and then merge it.

**Review by legoktm — CHANGES_REQUESTED** [review]
Unfortunately, I found a race condition, the `/var/lib/tor/authdir` directory is created by `anon-gw-anonymizer-config.service`, so we need a Requires/After (don't remember of the top of my head which one). Otherwise it'll error that the directory doesn't exist.

see https://github.com/Whonix/anon-gw-anonymizer-config/blob/master/etc/torrc.d/65_gateway.conf#L32-L39 and https://github.com/Whonix/anon-gw-anonymizer-config/blob/master/usr/libexec/anon-gw-anonymizer-config/make-sure-torrc-exist#L72.

Once that lands this should be ready to go, 🤞🏾

**Comment by cfm** [comment]
Thanks!  On it.

**Comment by cfm** [comment]
@legoktm in <https://github.com/freedomofpress/securedrop-workstation/pull/1051#pullrequestreview-2097483051>:

> Unfortunately, I found a race condition, the `/var/lib/tor/authdir` directory is created by `anon-gw-anonymizer-config.service`, so we need a Requires/After (don't remember of the top of my head which one). Otherwise it'll error that the directory doesn't exist.
> 
> see https://github.com/Whonix/anon-gw-anonymizer-config/blob/master/etc/torrc.d/65_gateway.conf#L32-L39 and https://github.com/Whonix/anon-gw-anonymizer-config/blob/master/usr/libexec/anon-gw-anonymizer-config/make-sure-torrc-exist#L72.
> 
> Once that lands this should be ready to go, 🤞🏾

Great catch: it looks like this was the cause of the original CI failure I wasn't able to reproduce locally.  Fixed proposed and explained (with one `FIXME` for follow-up) in freedomofpress/securedrop-client#2058.

**Review by legoktm — APPROVED** [review]
:shipit:

**Pull request merged**
